File: src/projection.js
/**
 * Modo projeção — tendência linear local amortecida sobre a média ponderada,
 * com horizonte limitado ao turno e bandas que alargam.
 *
 * Estimativa de modelo sobre a *média*, não pesquisa nova, não P(vitória),
 * não prognóstico de urna. Ver docs/modo-projecao-math.md.
 */

const DAY_MS = 86400000

/** 1º turno Eleições 2026 (TSE). */
export const ELECTION_ROUND1_MS = Date.parse('2026-10-04T12:00:00Z')
/** Eventual 2º turno. */
export const ELECTION_ROUND2_MS = Date.parse('2026-10-25T12:00:00Z')

export const PROJECTION_DEFAULTS = Object.freeze({
  fitDays: 14,
  horizonDays: 14,
  minPoints: 4,
  maxAbsSlope: 0.25, // pp / dia
  dampTauDays: 10,
  z: 1.645, // ~90%
  bandFloor: 2.0, // pp
  processSd: 0.12, // pp / √dia
})

/**
 * @param {{x:number,y:number}[]} series daily weighted-average points
 * @param {object} [opts]
 * @param {number} [opts.fitDays]
 * @param {number} [opts.horizonDays]
 * @param {number|null} [opts.electionDayMs] hard stop (inclusive day of vote)
 * @param {number} [opts.minPoints]
 * @param {number} [opts.maxAbsSlope]
 * @param {number} [opts.dampTauDays]
 * @param {number} [opts.z]
 * @param {number} [opts.bandFloor]
 * @param {number} [opts.processSd]
 * @returns {{
 *   ok: boolean,
 *   reason: string|null,
 *   line: {x:number,y:number}[],
 *   bandLow: {x:number,y:number}[],
 *   bandHigh: {x:number,y:number}[],
 *   lastObserved: number|null,
 *   slope: number|null,
 *   rmse: number|null,
 *   horizonUsed: number
 * }}
 */
export function projectTrend(series, opts = {}) {
  const fitDays = clamp(Math.round(opts.fitDays ?? PROJECTION_DEFAULTS.fitDays), 7, 28)
  const horizonCap = Math.max(1, Math.round(opts.horizonDays ?? PROJECTION_DEFAULTS.horizonDays))
  const minPoints = Math.max(2, opts.minPoints ?? PROJECTION_DEFAULTS.minPoints)
  const maxAbsSlope = opts.maxAbsSlope ?? PROJECTION_DEFAULTS.maxAbsSlope
  const dampTauDays = Math.max(1, opts.dampTauDays ?? PROJECTION_DEFAULTS.dampTauDays)
  const z = opts.z ?? PROJECTION_DEFAULTS.z
  const bandFloor = opts.bandFloor ?? PROJECTION_DEFAULTS.bandFloor
  const processSd = opts.processSd ?? PROJECTION_DEFAULTS.processSd
  const electionDayMs = opts.electionDayMs ?? null

  const empty = (reason) => ({
    ok: false,
    reason,
    line: [],
    bandLow: [],
    bandHigh: [],
    lastObserved: null,
    slope: null,
    rmse: null,
    horizonUsed: 0,
  })

  if (!series?.length) return empty('empty_series')

  const sorted = [...series].sort((a, b) => a.x - b.x)
  const tLast = sorted[sorted.length - 1].x
  const yLast = sorted[sorted.length - 1].y
  const tCut = tLast - fitDays * DAY_MS
  const window = sorted.filter((p) => p.x >= tCut)

  if (window.length < minPoints) {
    return { ...empty('insufficient_points'), lastObserved: tLast }
  }

  let daysToElection = Infinity
  if (electionDayMs != null && Number.isFinite(electionDayMs)) {
    daysToElection = Math.floor((electionDayMs - tLast) / DAY_MS)
    if (daysToElection < 1) {
      return { ...empty('past_or_on_election'), lastObserved: tLast }
    }
  }
  const horizonDays = Math.min(horizonCap, daysToElection === Infinity ? horizonCap : daysToElection)
  if (horizonDays < 1) {
    return { ...empty('horizon_zero'), lastObserved: tLast }
  }

  const t0 = window[0].x
  let sumT = 0
  let sumY = 0
  let sumTT = 0
  let sumTY = 0
  const n = window.length
  for (const p of window) {
    const t = (p.x - t0) / DAY_MS
    sumT += t
    sumY += p.y
    sumTT += t * t
    sumTY += t * p.y
  }
  const den = n * sumTT - sumT * sumT
  let b = den === 0 ? 0 : (n * sumTY - sumT * sumY) / den
  const a = (sumY - b * sumT) / n
  b = clip(b, -maxAbsSlope, maxAbsSlope)

  let sse = 0
  for (const p of window) {
    const t = (p.x - t0) / DAY_MS
    const e = p.y - (a + b * t)
    sse += e * e
  }
  const rmse = Math.sqrt(sse / Math.max(1, n - 2))

  const line = [{ x: tLast, y: round1(yLast) }]
  const bandLow = [{ x: tLast, y: round1(yLast) }]
  const bandHigh = [{ x: tLast, y: round1(yLast) }]

  for (let d = 1; d <= horizonDays; d++) {
    const x = tLast + d * DAY_MS
    // Tendência amortecida: Δ(d) = b * λ * (1 - e^{-d/λ})
    const delta = b * dampTauDays * (1 - Math.exp(-d / dampTauDays))
    const y = clip(yLast + delta, 0, 100)
    const widen =
      z *
      Math.sqrt(
        rmse * rmse * (1 + d / fitDays) + bandFloor * bandFloor + processSd * processSd * d,
      )
    line.push({ x, y: round1(y) })
    bandLow.push({ x, y: round1(clip(y - widen, 0, 100)) })
    bandHigh.push({ x, y: round1(clip(y + widen, 0, 100)) })
  }

  return {
    ok: true,
    reason: null,
    line,
    bandLow,
    bandHigh,
    lastObserved: tLast,
    slope: b,
    rmse,
    horizonUsed: horizonDays,
  }
}

/** Hex → rgba with alpha for band fill */
export function hexAlpha(hex, alpha) {
  const h = String(hex).replace('#', '')
  const full = h.length === 3 ? h.split('').map((c) => c + c).join('') : h
  const n = parseInt(full, 16)
  const r = (n >> 16) & 255
  const g = (n >> 8) & 255
  const b = n & 255
  return `rgba(${r},${g},${b},${alpha})`
}

/**
 * Copy pt-BR para o toggle / disclaimers (Frontend).
 * Manter strings estáveis para i18n futura.
 */
export const PROJECTION_COPY_PT = Object.freeze({
  toggleLabel: 'Modo projeção',
  toggleAria: 'Ativar extrapolação da tendência recente da média (estimativa de modelo)',
  chipOn: 'Projeção ligada',
  chartHint:
    'Linha tracejada = extrapolação amortecida da média ponderada recente, não uma pesquisa nova. A faixa indica incerteza do modelo (estimativa). Não é prognóstico de urna nem probabilidade de vitória.',
  methodology:
    'O Modo projeção estende a tendência recente da nossa média ponderada por poucos dias, com amortecimento e faixa de incerteza. É uma estimativa de modelo a partir de pesquisas já publicadas — não prevê o resultado da eleição e não substitui as pesquisas.',
  unavailable: 'Projeção indisponível: poucas pesquisas recentes nesta série.',
  seriesSuffix: 'projeção',
})

function clamp(v, lo, hi) {
  return Math.max(lo, Math.min(hi, v))
}

function clip(v, lo, hi) {
  return Math.max(lo, Math.min(hi, v))
}

function round1(v) {
  return Math.round(v * 10) / 10
}

File: rust/polling-core/src/projection.rs
use crate::{DAY_MS, SeriesPoint};

pub const ELECTION_ROUND1_MS: f64 = 1_791_100_800_000.0;
pub const ELECTION_ROUND2_MS: f64 = 1_792_912_800_000.0;

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct ProjectPoint {
    pub x: f64,
    pub y: f64,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ProjectionResult {
    pub ok: bool,
    pub reason: Option<String>,
    pub line: Vec<ProjectPoint>,
    pub band_low: Vec<ProjectPoint>,
    pub band_high: Vec<ProjectPoint>,
    pub last_observed: Option<f64>,
    pub slope: Option<f64>,
    pub rmse: Option<f64>,
    pub horizon_used: u32,
}

fn empty(reason: &str) -> ProjectionResult {
    ProjectionResult {
        ok: false,
        reason: Some(reason.to_string()),
        line: Vec::new(),
        band_low: Vec::new(),
        band_high: Vec::new(),
        last_observed: None,
        slope: None,
        rmse: None,
        horizon_used: 0,
    }
}

fn clamp(value: f64, low: f64, high: f64) -> f64 {
    value.max(low).min(high)
}

fn round1(value: f64) -> f64 {
    (value * 10.0).round() / 10.0
}

pub fn project_trend(series: &[SeriesPoint], fit_days: f64, horizon_days: u32, election_day_ms: Option<f64>, min_points: usize, max_abs_slope: f64, damp_tau_days: f64, z: f64, band_floor: f64, process_sd: f64) -> ProjectionResult {
    if series.is_empty() { return empty("empty_series"); }

    let fit_days = if fit_days.is_finite() { fit_days.round().clamp(7.0, 28.0) } else { 14.0 };
    let horizon_cap = horizon_days.max(1);
    let min_points = min_points.max(2);
    let max_abs_slope = if max_abs_slope.is_finite() { max_abs_slope } else { 0.25 };
    let damp_tau_days = if damp_tau_days.is_finite() { damp_tau_days.max(1.0) } else { 10.0 };
    let z = if z.is_finite() { z } else { 1.645 };
    let band_floor = if band_floor.is_finite() { band_floor } else { 2.0 };
    let process_sd = if process_sd.is_finite() { process_sd } else { 0.12 };

    let mut sorted = series.to_vec();
    sorted.sort_by(|a, b| a.x.total_cmp(&b.x));
    let t_last = sorted.last().unwrap().x;
    let y_last = sorted.last().unwrap().y;
    let t_cut = t_last - fit_days * DAY_MS;
    let window: Vec<SeriesPoint> = sorted.into_iter().filter(|p| p.x >= t_cut).collect();
    if window.len() < min_points {
        let mut result = empty("insufficient_points");
        result.last_observed = Some(t_last);
        return result;
    }

    let mut days_to_election = f64::INFINITY;
    if let Some(election) = election_day_ms.filter(|v| v.is_finite()) {
        days_to_election = ((election - t_last) / DAY_MS).floor();
        if days_to_election < 1.0 {
            let mut result = empty("past_or_on_election");
            result.last_observed = Some(t_last);
            return result;
        }
    }
    let horizon = (horizon_cap as f64).min(days_to_election) as u32;
    if horizon < 1 {
        let mut result = empty("horizon_zero");
        result.last_observed = Some(t_last);
        return result;
    }

    let t0 = window[0].x;
    let mut sum_t = 0.0;
    let mut sum_y = 0.0;
    let mut sum_tt = 0.0;
    let mut sum_ty = 0.0;
    let n = window.len() as f64;
    for p in &window {
        let t = (p.x - t0) / DAY_MS;
        sum_t += t;
        sum_y += p.y;
        sum_tt += t * t;
        sum_ty += t * p.y;
    }
    let denominator = n * sum_tt - sum_t * sum_t;
    let mut slope = if denominator == 0.0 { 0.0 } else { (n * sum_ty - sum_t * sum_y) / denominator };
    let intercept = (sum_y - slope * sum_t) / n;
    slope = clamp(slope, -max_abs_slope, max_abs_slope);

    let mut sse = 0.0;
    for p in &window {
        let t = (p.x - t0) / DAY_MS;
        let e = p.y - (intercept + slope * t);
        sse += e * e;
    }
    let rmse = (sse / (window.len().saturating_sub(2).max(1) as f64)).sqrt();

    let mut line = vec![ProjectPoint { x: t_last, y: round1(y_last) }];
    let mut band_low = vec![ProjectPoint { x: t_last, y: round1(y_last) }];
    let mut band_high = vec![ProjectPoint { x: t_last, y: round1(y_last) }];

    for day in 1..=horizon {
        let x = t_last + day as f64 * DAY_MS;
        let d = day as f64;
        let delta = slope * damp_tau_days * (1.0 - (-d / damp_tau_days).exp());
        let y = clamp(y_last + delta, 0.0, 100.0);
        let widen = z * (rmse * rmse * (1.0 + d / fit_days) + band_floor * band_floor + process_sd * process_sd * d).sqrt();
        line.push(ProjectPoint { x, y: round1(y) });
        band_low.push(ProjectPoint { x, y: round1(clamp(y - widen, 0.0, 100.0)) });
        band_high.push(ProjectPoint { x, y: round1(clamp(y + widen, 0.0, 100.0)) });
    }

    ProjectionResult {
        ok: true,
        reason: None,
        line,
        band_low,
        band_high,
        last_observed: Some(t_last),
        slope: Some(slope),
        rmse: Some(rmse),
        horizon_used: horizon,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rejects_empty_series() {
        assert_eq!(project_trend(&[], 14.0, 14, None, 4, 0.25, 10.0, 1.645, 2.0, 0.12).reason.as_deref(), Some("empty_series"));
    }

    #[test]
    fn caps_slope_and_preserves_current_point() {
        let rows: Vec<SeriesPoint> = (0..6).map(|day| SeriesPoint { x: day as f64 * DAY_MS, y: 40.0 + day as f64 }).collect();
        let result = project_trend(&rows, 14.0, 4, None, 4, 0.25, 10.0, 1.645, 2.0, 0.12);
        assert!(result.ok);
        assert!((result.slope.unwrap() - 0.25).abs() < 1e-12);
        assert_eq!(result.line[0].x, 5.0 * DAY_MS);
        assert_eq!(result.line[0].y, 45.0);
        assert_eq!(result.horizon_used, 4);
    }
}
